### PR TITLE
TEZ-4527: Add generic and pluggable hooks for DAGs and task attempts

### PR DESCRIPTION
**Pull request by okumin**
https://issues.apache.org/jira/browse/TEZ-4527

### CODE REVIEW

**Comment by tez-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  26m 20s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 24s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  12m 38s |  master passed  |
| +1 :green_heart: |  compile  |   2m 13s |  master passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  compile  |   2m  2s |  master passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  checkstyle  |   2m  3s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 52s |  master passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   1m 39s |  master passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +0 :ok: |  spotbugs  |   1m 18s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 14s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m  9s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 19s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 27s |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javac  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 16s |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  javac  |   1m 16s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m  8s |  tez-common: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| -0 :warning: |  checkstyle  |   0m 20s |  tez-dag: The patch generated 1 new + 51 unchanged - 0 fixed = 52 total (was 51)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  findbugs  |   3m 31s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 14s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 26s |  tez-common in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 38s |  tez-runtime-internals in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 56s |  tez-dag in the patch passed.  |
| -1 :x: |  asflicense  |   0m 38s |  The patch generated 4 ASF License warnings.  |
|  |   |  80m 23s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/324 |
| JIRA Issue | TEZ-4527 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux 3c822e51fa12 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 0c5cf6886 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~22.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~22.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/artifact/out/diff-checkstyle-tez-common.txt |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/artifact/out/diff-checkstyle-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 416 (vs. ulimit of 5500) |
| modules | C: tez-api tez-common tez-runtime-internals tez-dag U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/1/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java [hunk 1]
```diff
@@ -2297,12 +2297,14 @@ static Set<String> getPropertySet() {
...
-   *  Frequency at which thread dump should be captured. Supports TimeUnits.
+   *  Frequency at which thread dump should be captured. Supports TimeUnits. This is effective only
+   *  when org.apache.tez.dag.app.ThreadDumpDAGHook is configured to tez.am.hooks or
+   *  org.apache.tez.runtime.task.ThreadDumpTaskAttemptHook is configured to tez.task.attempt.hooks.
    */
   @ConfigurationScope(Scope.DAG)
   @ConfigurationProperty
   public static final String TEZ_THREAD_DUMP_INTERVAL = "tez.thread.dump.interval";
-  public static final String TEZ_THREAD_DUMP_INTERVAL_DEFAULT = "0ms";
+  public static final String TEZ_THREAD_DUMP_INTERVAL_DEFAULT = "100ms";
```
Comment (okumin) [inline tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java:2307]: If we introduce pluggable hooks, I think we can change the default value. We may remove NOOP_TEZ_THREAD_DUMP_HELPER, too.

**Comment by tez-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 15s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  10m 24s |  master passed  |
| +1 :green_heart: |  compile  |   2m 14s |  master passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  compile  |   2m  2s |  master passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  checkstyle  |   1m 56s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 49s |  master passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   1m 39s |  master passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +0 :ok: |  spotbugs  |   1m 22s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 19s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 19s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 25s |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javac  |   1m 25s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 15s |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  javac  |   1m 15s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m  9s |  tez-common: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| -0 :warning: |  checkstyle  |   0m 20s |  tez-dag: The patch generated 1 new + 51 unchanged - 0 fixed = 52 total (was 51)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~22.04-b08  |
| +1 :green_heart: |  findbugs  |   3m 32s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 14s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 26s |  tez-common in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 37s |  tez-runtime-internals in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 56s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 39s |  The patch does not generate ASF License warnings.  |
|  |   |  52m  5s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/324 |
| JIRA Issue | TEZ-4527 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux ad0fa2a80cb4 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 0c5cf6886 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~22.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu122.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~22.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/2/artifact/out/diff-checkstyle-tez-common.txt |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/2/artifact/out/diff-checkstyle-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/2/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: tez-api tez-common tez-runtime-internals tez-dag U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/2/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java (hunk 1 above)
Comment (abstractdog) [reply tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java:2307]: makes sense
I think for package clarity's sake, all the hook related configs can go a namespace that implies they're hooks:
```
tez.hook.thread.dump.internal
```
also:
```
TEZ_HOOK_THREAD_DUMP_INTERVAL
```

File: tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java [hunk 2]
```diff
@@ -2207,7 +2210,9 @@ public Void run() throws Exception {
     }
 
     // Check if the thread dump service is up in any case, if yes attempt a shutdown
```
Comment (abstractdog) [inline tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java:2212]: remove thread dump helper related comment, and change to a more generic one that tells we're about to stop hooks if they are running in any case

File: tez-runtime-internals/src/main/java/org/apache/tez/runtime/task/ThreadDumpTaskAttemptHook.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/**
```
Comment (abstractdog) [inline tez-runtime-internals/src/main/java/org/apache/tez/runtime/task/ThreadDumpTaskAttemptHook.java:1]: shouldn't this go to the new package: org.apache.tez.runtime.hook

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java (hunk 1 above)
Comment (abstractdog) [reply tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java:2307]: I think you can remove the NoopTezThreadDumpHelper

**Comment by abstractdog** [comment]
@okumin : thanks for the patch, nice refactor, only minor comments, other than that, it looks good to me!

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java (hunk 1 above)
Comment (okumin) [reply tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java]: Done.
- https://github.com/apache/tez/pull/324/commits/1673bd5a88f94b097ca3c8cb627553e30efb6b6c
- https://github.com/apache/tez/pull/324/commits/79fe8d036a3fe7cae0a66163a7a8d983ffa3a14e

File: tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java (hunk 2 above)
Comment (okumin) [reply tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java]: Done
https://github.com/apache/tez/pull/324/commits/943012ab3aeefcfe2e154a669e45c0999c366cb1

**Comment by okumin** [comment]
Thanks. I think all the points follow your suggestions. I rebased the branch as the original one was already obsolete

**Comment by tez-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  25m 59s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   2m 53s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  13m 33s |  master passed  |
| +1 :green_heart: |  compile  |   2m 53s |  master passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  compile  |   2m 37s |  master passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  checkstyle  |   2m 38s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 21s |  master passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   2m  6s |  master passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +0 :ok: |  spotbugs  |   0m 50s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 11s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 43s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 51s |  the patch passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javac  |   1m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 34s |  the patch passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  javac  |   1m 34s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 10s |  tez-runtime-internals: The patch generated 1 new + 7 unchanged - 0 fixed = 8 total (was 7)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   1m  2s |  the patch passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   1m  1s |  the patch passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  findbugs  |   4m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 15s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 26s |  tez-common in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 47s |  tez-runtime-internals in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 49s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  unit  |  40m 54s |  tez-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 55s |  The patch does not generate ASF License warnings.  |
|  |   | 125m  6s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/324 |
| JIRA Issue | TEZ-4527 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux cebf547baff1 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / ca1511998 |
| Default Java | Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/4/artifact/out/diff-checkstyle-tez-runtime-internals.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/4/testReport/ |
| Max. process+thread count | 1203 (vs. ulimit of 5500) |
| modules | C: tez-api tez-common tez-runtime-internals tez-dag tez-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/4/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java [hunk 4]
```diff
@@ -70,21 +71,17 @@ private TezThreadDumpHelper(long duration, Configuration conf) throws IOExceptio
         "path: {}", duration, basePath);
   }
 
-  public TezThreadDumpHelper() {
```
Comment (abstractdog) [inline tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java:73]: hm, cannot recall what the purpose was of this constructor, does reflection work without this explicitly defined?
I'm afraid that as there is private parameterized constructor, class.newInstance() throws an InstantiationException, doesn't it?

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java [hunk 5]
```diff
@@ -2297,12 +2297,14 @@ static Set<String> getPropertySet() {
...
-   *  Frequency at which thread dump should be captured. Supports TimeUnits.
+   *  Frequency at which thread dump should be captured. Supports TimeUnits. This is effective only
+   *  when org.apache.tez.dag.app.ThreadDumpDAGHook is configured to tez.am.hooks or
+   *  org.apache.tez.runtime.task.ThreadDumpTaskAttemptHook is configured to tez.task.attempt.hooks.
    */
   @ConfigurationScope(Scope.DAG)
   @ConfigurationProperty
-  public static final String TEZ_THREAD_DUMP_INTERVAL = "tez.thread.dump.interval";
-  public static final String TEZ_THREAD_DUMP_INTERVAL_DEFAULT = "0ms";
+  public static final String TEZ_HOOK_THREAD_DUMP_INTERVAL = "tez.hook.thread.dump.interval";
```
Comment (abstractdog) [inline tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java:2306]: @okumin : I'm terribly sorry, I just realized that changing this causes more problems than benefits (changing config opts from one release to another), the class name also doesn't have "hook" in it, so it's fine to have this as "tez.thread.dump.interval", are you fine with chaging back?

**Comment by abstractdog** [comment]
> Thanks. I think all the points follow your suggestions. I rebased the branch as the original one was already obsolete

thanks @okumin , this is very close, just left 2 comments

File: tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java (hunk 5 above)
Comment (okumin) [reply tez-api/src/main/java/org/apache/tez/dag/api/TezConfiguration.java]: np. I renamed them back
https://github.com/apache/tez/pull/324/commits/293ce63e334e732c742cdc29ed6fe0fd9635d097

File: tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java (hunk 4 above)
Comment (okumin) [reply tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java:73]: It is originally needed to instantiate [https://github.com/apache/tez/blob/ca1511998e183c8e10da79bef20463d40a39726a/tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java#L182-L194](https://github.com/apache/tez/blob/ca1511998e183c8e10da79bef20463d40a39726a/tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java#L182-L194) with zero arguments.

I think the class is not constructed in a reflective way, or it doesn't assume it's reflectively operated. I slightly updated the modifiers to make sure it
https://github.com/apache/tez/pull/324/commits/61d8249fc9051acc1b98dfe6059ea00c4634e6ce

Comment (abstractdog) [reply tez-runtime-internals/src/main/java/org/apache/tez/runtime/TezThreadDumpHelper.java:73]: right, I was wrong, the hooks are created by reflection, the TezThreadDumpHelper is not
```
    helper = TezThreadDumpHelper.getInstance(conf).start(id.toString());
```

**Review by abstractdog — APPROVED** [review]
LGTM +1

**Comment by tez-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   2m 53s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  13m 20s |  master passed  |
| +1 :green_heart: |  compile  |   2m 53s |  master passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  compile  |   2m 37s |  master passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  checkstyle  |   2m 38s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 24s |  master passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   2m  6s |  master passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +0 :ok: |  spotbugs  |   0m 47s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m  7s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 40s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 48s |  the patch passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javac  |   1m 48s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 36s |  the patch passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  javac  |   1m 36s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 10s |  tez-runtime-internals: The patch generated 1 new + 7 unchanged - 0 fixed = 8 total (was 7)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   1m  3s |  the patch passed with JDK Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04  |
| +1 :green_heart: |  javadoc  |   1m  2s |  the patch passed with JDK Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga  |
| +1 :green_heart: |  findbugs  |   4m 14s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 14s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 25s |  tez-common in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 48s |  tez-runtime-internals in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 49s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  unit  |  42m 22s |  tez-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  The patch does not generate ASF License warnings.  |
|  |   | 100m 52s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/324 |
| JIRA Issue | TEZ-4527 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux 972ccfe046fc 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / ca1511998 |
| Default Java | Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.25+9-post-Ubuntu-1ubuntu122.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_432-8u432-ga~us1-0ubuntu2~22.04-ga |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/5/artifact/out/diff-checkstyle-tez-runtime-internals.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/5/testReport/ |
| Max. process+thread count | 1173 (vs. ulimit of 5500) |
| modules | C: tez-api tez-common tez-runtime-internals tez-dag tez-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-324/5/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by okumin** [comment]
Thank you. This change is so helpful for us

**Pull request merged**